src/connectors/impls/s3/streamer.rs

```diff
@@ -160,51 +181,47 @@ impl Sink for S3Sink {
         _start: u64,
     ) -> Result<SinkReply> {
         let ingest_id = event.ingest_ns;
-
-        for (event, meta) in event.value_meta_iter() {
-            // Handle no key in meta.
-
-            let s3_meta = S3Meta::new(meta);
+        let mut event_tracked = false;
+        for (value, meta) in event.value_meta_iter() {
+            let s3_meta = S3Meta::new(ctx.extract_meta(meta));
 
             let object_key = if let Some(key) = s3_meta.get_object_key().map(ToString::to_string) {
                 key
             } else {
-                self.current_key.clear();
-                error!("{ctx}: missing '$s3_streamer.key' meta data in event");
+                // Handle no key in meta.
+                error!("{ctx}: missing '${CONNECTOR_TYPE}.key' meta data in event");
                 return Ok(SinkReply::FAIL);
             };
 
             if object_key != self.current_key {
                 // we switched keys:
                 // 1. finish the current upload, if any
-                // 2. initiate a new upload
-                self.prepare_new_multipart(object_key, ctx).await?;
+                // 2. initiate a new upload and start trackign the new event
+                self.prepare_new_multipart(object_key, &event, ctx).await?;
+            } else if !event_tracked {
+                // track event
+                self.current_event_id.track(&event.id);
+                if !event.op_meta.is_empty() {
+                    self.current_op_meta.merge(event.op_meta.clone());
+                }
+                self.current_transactional |= event.transactional;
+                event_tracked = true;
             }
 
             // Handle the aggregation.
-            for data in serializer.serialize(event, ingest_id)? {
+            for data in serializer.serialize(value, ingest_id)? {
```

Comment:
we could re-use the buffer from gcl here - unify the behaviour that would be nice :)

Reply:
The gcs buffer is more complicated than it needs to be, because gcs decided to not guarantee that all bytes in a request have been written, so we need to track what has been committed by the backend separately. I would keep this one simple and keep using a `Vec<u8>` here.